# Evolution of Beyonce's Album

For my final project, I will be exploring the evolution of Beyonce's albums by analyzing the trend of sentiment scores. The data used in this project includes the names of albums, titles of song, song lyrics and release year of album (link: https://www.kaggle.com/datasets/hillaryosei/beyonce-lyrics?resource=download). 
<br>Another dataset, which I will be making myself, simply shows the number of copies sold for each album (raw data: https://www.forbes.com/sites/marisadellatto/2022/08/08/beyoncs-renaissance-is-a-no-1-album-but-its-not-her-biggest-debut-ever/?sh=599029c227d5)

My motivation for selecting this dataset is two-fold: a) Beyonce is one of the most legendary artists of the century with her career spanning over 20 years. Therefore, I believe that she would be an interesting topic of a project. b) Throughout the DH 140 course, I particularly enjoyed text analysis and data visualization. Thus, I want to use these two techniques primarily. With these reasons in mind, my research question is **this:**

***How have the moods of Beyonce's albums transformed over the span of her career and what is the correlation between moods and sales of her album?***


### I. Exploring Data 
<br> Let's first explore the Beyonce dataset that I found from online. 

In [128]:
import pandas as pd
import matplotlib.pyplot as plt
beyonce_file=pd.read_csv("beyonce_lyrics - beyonce_lyrics(1).csv")
beyonce_file



artist                album         track_title  track_num  \
0     Beyoncé  Dangerously in Love       Crazy in Love          1   
1     Beyoncé  Dangerously in Love       Crazy in Love          1   
2     Beyoncé  Dangerously in Love       Crazy in Love          1   
3     Beyoncé  Dangerously in Love       Crazy in Love          1   
4     Beyoncé  Dangerously in Love       Crazy in Love          1   
...       ...                  ...                 ...        ...   
6151  Beyoncé          RENAISSANCE  SUMMER RENAISSANCE         16   
6152  Beyoncé          RENAISSANCE  SUMMER RENAISSANCE         16   
6153  Beyoncé          RENAISSANCE  SUMMER RENAISSANCE         16   
6154  Beyoncé          RENAISSANCE  SUMMER RENAISSANCE         16   
6155  Beyoncé          RENAISSANCE  SUMMER RENAISSANCE         16   

                                                  lyric  line  year  
0                               Yes, so crazy right now     1  2003  
1                     Most incredibly, it's your girl B     2  2003  
2                                  It's your boy, young     3  2003  
3                          History in the makin' part 2     4  2003  
4                                    So crazy right now     5  2003  
...                                                 ...   ...   ...  
6151  This Telfar bag imported, Birkins, them shits ...    63  2022  
6152                                      I'm in my bag    64  2022  
6153                                             Ah-ooh    65  2022  
6154                                             Ah-ooh    66  2022  
6155                                             Ah-ooh    67  2022  

[6156 rows x 7 columns]

In [17]:
#drop the column called artist
beyonce_file.drop('artist', inplace=True, axis=1)
beyonce_file

album         track_title  track_num  \
0     Dangerously in Love       Crazy in Love          1   
1     Dangerously in Love       Crazy in Love          1   
2     Dangerously in Love       Crazy in Love          1   
3     Dangerously in Love       Crazy in Love          1   
4     Dangerously in Love       Crazy in Love          1   
...                   ...                 ...        ...   
6151          RENAISSANCE  SUMMER RENAISSANCE         16   
6152          RENAISSANCE  SUMMER RENAISSANCE         16   
6153          RENAISSANCE  SUMMER RENAISSANCE         16   
6154          RENAISSANCE  SUMMER RENAISSANCE         16   
6155          RENAISSANCE  SUMMER RENAISSANCE         16   

                                                  lyric  line  year  
0                               Yes, so crazy right now     1  2003  
1                     Most incredibly, it's your girl B     2  2003  
2                                  It's your boy, young     3  2003  
3                          History in the makin' part 2     4  2003  
4                                    So crazy right now     5  2003  
...                                                 ...   ...   ...  
6151  This Telfar bag imported, Birkins, them shits ...    63  2022  
6152                                      I'm in my bag    64  2022  
6153                                             Ah-ooh    65  2022  
6154                                             Ah-ooh    66  2022  
6155                                             Ah-ooh    67  2022  

[6156 rows x 6 columns]

Further pre-analyzing the dataset, you can see below that the average number of tracks in each album is 7.5, the average number of lines per each track is 37. 

In [33]:
beyonce_file.loc[:,["track_num", "line"]].describe()

track_num         line
count  6156.000000  6156.000000
mean      7.462801    36.922840
std       4.142908    24.479552
min       1.000000     1.000000
25%       4.000000    17.000000
50%       7.000000    34.000000
75%      11.000000    54.000000
max      16.000000   122.000000

Next, I want to present how many tracks each album has. Since each album was released in different year, I can simply distinguish album based on the year column. Below, I created a small dataset that only has the track_num and year columns:

In [96]:
beyonce_file.loc[:,['track_num','year']]

track_num  year
0             1  2003
1             1  2003
2             1  2003
3             1  2003
4             1  2003
...         ...   ...
6151         16  2022
6152         16  2022
6153         16  2022
6154         16  2022
6155         16  2022

[6156 rows x 2 columns]

However, since there are 6156 rows, it is impossible to clearly see the number of tracks in each album. Thus, I decided to create a variable for each album and with that, it would be very easy to quickly grasp the result. First, let's get the list of all the albums: 

In [120]:
print(beyonce_file['album'].drop_duplicates())

0       Dangerously in Love
914                   B'Day
1884    I Am...Sasha Fierce
2495                      4
3451                BEYONCÉ
4349               Lemonade
4966            RENAISSANCE
Name: album, dtype: object


Below are the basic info of each album. I will disregard 'track_title', 'lyric' and 'line' now and just focus on 'album', 'track_num' and 'year'.

In [121]:
dangerously_in_love=beyonce_file.loc[beyonce_file['album']=='Dangerously in Love']
dangerously_in_love.loc[dangerously_in_love['track_num'].idxmax()]



album                                        Dangerously in Love
track_title                                                Daddy
track_num                                                     15
lyric          I remember when you use to take me on a bike e...
line                                                           1
year                                                        2003
Name: 862, dtype: object

In [122]:
bday=beyonce_file.loc[beyonce_file['album']=="B'Day"]
bday.loc[bday['track_num'].idxmax()]


album                                        B'Day
track_title                                 Listen
track_num                                       13
lyric          Listen to the song here in my heart
line                                             1
year                                          2006
Name: 1837, dtype: object

In [123]:
sasha=beyonce_file.loc[beyonce_file['album']=='I Am...Sasha Fierce']
sasha.loc[sasha['track_num'].idxmax()]


album                 I Am...Sasha Fierce
track_title                   Video Phone
track_num                              11
lyric          Shawty, what your name is?
line                                    1
year                                 2008
Name: 2423, dtype: object

In [124]:
four=beyonce_file.loc[beyonce_file['album']=='4']
four.loc[four['track_num'].idxmax()]


album                                                          4
track_title                                        Dance For You
track_num                                                     16
lyric          I just wanna show you how much I appreciate yo...
line                                                           1
year                                                        2011
Name: 3362, dtype: object

In [125]:
beyonce=beyonce_file.loc[beyonce_file['album']=='BEYONCÉ']
beyonce.loc[beyonce['track_num'].idxmax()]

album                                               BEYONCÉ
track_title                                            Blue
track_num                                                14
lyric          Sometimes, these walls seem to cave in on me
line                                                      1
year                                                   2013
Name: 4306, dtype: object

In [126]:
lemonade=beyonce_file.loc[beyonce_file['album']=='Lemonade']
lemonade.loc[lemonade['track_num'].idxmax()]



album                                   Lemonade
track_title                            Formation
track_num                                     12
lyric          What happened at the New Wil'ins?
line                                           1
year                                        2018
Name: 4897, dtype: object

In [127]:
renaissance=beyonce_file.loc[beyonce_file['year']==2022]
renaissance.loc[renaissance['track_num'].idxmax()]

album                                          RENAISSANCE
track_title                             SUMMER RENAISSANCE
track_num                                               16
lyric          I wanna house you and make you take my name
line                                                     1
year                                                  2022
Name: 6089, dtype: object

In [108]:
#print all the lyrics of an album


In [129]:
#tokenize 

In [ ]:
#eliminate stopwords, punctuations, porter stemmer

In [ ]:
#print 30 most frequent words in each album

In [ ]:
#graph the words with frequency

In [ ]:
#calculate sentiment score for each album
# vadar 


In [ ]:
#compile all scores into 1 complete dataframe

In [ ]:
#graph it 

In [ ]:
#make a graph of the sales

In [ ]:
#compare the shape and comment 

In [ ]:
#conclusion 